### ISSUE TITLE
[r2.14.0, rtsp library] setting repeat mode causes fatal error, resulting in force close

### ISSUE THREAD

**Issue by warren-bank** [google/ExoPlayer#8994]
<details>
  <summary>logcat errors:</summary>
  
  ```text
    05-29 12:15:00.692 26748-26882/com.github.warren_bank.exoplayer_airplay_receiver E/AndroidRuntime: FATAL EXCEPTION: ExoPlayer:Playback
        Process: com.github.warren_bank.exoplayer_airplay_receiver, PID: 26748
        java.lang.NullPointerException
            at com.google.android.exoplayer2.util.Assertions.checkNotNull(Assertions.java:154)
            at com.google.android.exoplayer2.source.rtsp.RtpDataLoadable.setTimestamp(RtpDataLoadable.java:111)
            at com.google.android.exoplayer2.source.rtsp.RtspMediaPeriod$InternalListener.onPlaybackStarted(RtspMediaPeriod.java:495)
            at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener.onPlayResponseReceived(RtspClient.java:496)
            at com.google.android.exoplayer2.source.rtsp.RtspClient$MessageListener.onRtspMessageReceived(RtspClient.java:412)
            at com.google.android.exoplayer2.source.rtsp.RtspMessageChannel$Receiver.lambda$handleRtspMessage$0(RtspMessageChannel.java:291)
            at com.google.android.exoplayer2.source.rtsp.-$$Lambda$RtspMessageChannel$Receiver$frEfu1WfqENR1_7bFfrua9iV9BY.run(lambda)
            at android.os.Handler.handleCallback(Handler.java:733)
            at android.os.Handler.dispatchMessage(Handler.java:95)
            at android.os.Looper.loop(Looper.java:136)
            at android.os.HandlerThread.run(HandlerThread.java:61)
    05-29 12:15:03.062 26748-26748/com.github.warren_bank.exoplayer_airplay_receiver E/EventLogger: playerFailed [eventTime=219.65, mediaPos=0.00, window=0, period=0
          com.google.android.exoplayer2.ExoPlaybackException: null error, index=-1, format=null, format_supported=YES
            at com.google.android.exoplayer2.SimpleExoPlayer.setVideoOutputInternal(SimpleExoPlayer.java:1922)
            at com.google.android.exoplayer2.SimpleExoPlayer.access$4000(SimpleExoPlayer.java:92)
            at com.google.android.exoplayer2.SimpleExoPlayer$ComponentListener.surfaceDestroyed(SimpleExoPlayer.java:2278)
            at android.view.SurfaceView.updateWindow(SurfaceView.java:556)
            at android.view.SurfaceView.access$000(SurfaceView.java:86)
            at android.view.SurfaceView$1.handleMessage(SurfaceView.java:129)
            at android.os.Handler.dispatchMessage(Handler.java:102)
            at android.os.Looper.loop(Looper.java:136)
            at android.app.ActivityThread.main(ActivityThread.java:5118)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:515)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:790)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:606)
            at dalvik.system.NativeStart.main(Native Method)
         Caused by: com.google.android.exoplayer2.ExoTimeoutException: Detaching surface timed out.
            at com.google.android.exoplayer2.SimpleExoPlayer.setVideoOutputInternal(SimpleExoPlayer.java:1920)
            at com.google.android.exoplayer2.SimpleExoPlayer.access$4000(SimpleExoPlayer.java:92) 
            at com.google.android.exoplayer2.SimpleExoPlayer$ComponentListener.surfaceDestroyed(SimpleExoPlayer.java:2278) 
            at android.view.SurfaceView.updateWindow(SurfaceView.java:556) 
            at android.view.SurfaceView.access$000(SurfaceView.java:86) 
            at android.view.SurfaceView$1.handleMessage(SurfaceView.java:129) 
            at android.os.Handler.dispatchMessage(Handler.java:102) 
            at android.os.Looper.loop(Looper.java:136) 
            at android.app.ActivityThread.main(ActivityThread.java:5118) 
            at java.lang.reflect.Method.invokeNative(Native Method) 
            at java.lang.reflect.Method.invoke(Method.java:515) 
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:790) 
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:606) 
            at dalvik.system.NativeStart.main(Native Method) 
        ]
  ```
</details>

- - - -

__quick trace through code to find cause:__

* method: [RtpDataLoadable.setTimestamp](https://github.com/google/ExoPlayer/blob/r2.14.0/library/rtsp/src/main/java/com/google/android/exoplayer2/source/rtsp/RtpDataLoadable.java#L109)
  ```java
    public void setTimestamp(long timestamp) {
      if (!checkNotNull(extractor).hasReadFirstRtpPacket()) {...}
    }
   ```
* method: [RtpDataLoadable.load](https://github.com/google/ExoPlayer/blob/r2.14.0/library/rtsp/src/main/java/com/google/android/exoplayer2/source/rtsp/RtpDataLoadable.java#L148)
  ```java
    public void load() {
      extractor = new RtpExtractor(rtspMediaTrack.payloadFormat, trackId);
    }
  ```
* inner class constructor: [RtspMediaPeriod.RtpLoadInfo](https://github.com/google/ExoPlayer/blob/r2.14.0/library/rtsp/src/main/java/com/google/android/exoplayer2/source/rtsp/RtspMediaPeriod.java#L684)
  ```java
    public RtpLoadInfo(...){
      this.loadable = new RtpDataLoadable(...);
    }
  ```
* method: [RtspMediaPeriod.getLoadableByTrackUri](https://github.com/google/ExoPlayer/blob/r2.14.0/library/rtsp/src/main/java/com/google/android/exoplayer2/source/rtsp/RtspMediaPeriod.java#L307)
  ```java
    private RtpDataLoadable getLoadableByTrackUri(Uri trackUri) {
      RtpLoadInfo loadInfo = rtspLoaderWrappers.get(i).loadInfo;
      return loadInfo.loadable;
    }
  ```
* method: [RtspMediaPeriod.onPlaybackStarted](https://github.com/google/ExoPlayer/blob/r2.14.0/library/rtsp/src/main/java/com/google/android/exoplayer2/source/rtsp/RtspMediaPeriod.java#L495)
  ```java
    public void onPlaybackStarted(...) {
      RtpDataLoadable dataLoadable = getLoadableByTrackUri(trackTiming.uri);

      // NEVER CALLS: dataLoadable.load();

      dataLoadable.setTimestamp(trackTiming.rtpTimestamp);
      // ERROR: dataLoadable.extractor is NULL
    }
  ```

**Comment by claincly** [comment]
Hello there, can you also post the RTSP messages? You can print the messages by following this simple step
https://github.com/google/ExoPlayer/pull/3854#issuecomment-841167938

**Comment by warren-bank** [comment]
I must be slow.. that doesn't look simple to me :frowning_face: 

[Here](https://github.com/warren-bank/Android-ExoPlayer-AirPlay-Receiver/blob/v2.0.22/android-studio-project/ExoPlayer-AirPlay-Receiver/src/main/java/com/github/warren_bank/exoplayer_airplay_receiver/exoplayer2/PlayerManager.java#L997) is where the `RtspMediaSource` is initialized. The following code removes all of the app's complexity and summarizes how it is used:

```java
  String uri = "rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_175k.mov";
  String mimeType = "application/x-rtsp";
  MediaItem mediaItem = new MediaItem.Builder().setUri(uri).setMimeType(mimeType).build();
  MediaSource mediaSource = (MediaSource) new RtspMediaSource.Factory().createMediaSource(mediaItem);

  ConcatenatingMediaSource concatenatingMediaSource = new ConcatenatingMediaSource();
  MediaSource[] mediaSources = new MediaSource[]{mediaSource};
  concatenatingMediaSource.addMediaSources(
    Arrays.asList(mediaSources)
  );

  SimpleExoPlayer player = new SimpleExoPlayer.Builder(context).build();
  player.prepare(concatenatingMediaSource);
```

Is there a way to hook the `AnalyticsCollector` (ie: configure an `AnalyticsListener`) to the RTSP library.. so it can provide a deep log into all RTSP-related events? I'm currently using the stock [EventLogger](https://github.com/google/ExoPlayer/blob/r2.14.0/library/core/src/main/java/com/google/android/exoplayer2/util/EventLogger.java).

**Comment by warren-bank** [comment]
```text
  - needs triage
  + need more info
```

* I made a minimal barebones app to run the code in the previous comment, and (to my surprise) it works correctly
  - [code](https://github.com/warren-bank/Android-ExoPlayer-demos/blob/r2.14.0.000/4-rtsp-barebones/src/main/java/com/google/android/exoplayer2/rtspdemo/MainActivity.java#L89)
  - [apk](https://github.com/warren-bank/Android-ExoPlayer-demos/releases/tag/r2.14.0.000)
* so I need to temporarily put a pin in this issue
  - run the app that's experiencing a problem through the debugger
  - try to figure out why it crashes (but only when loading an RTSP url)

**Comment by claincly** [comment]
Sorry I meant the instructions to add the logs to see the actual exchanged RTSP messages.

To start, could you try plugging your link to any of RTSP links in 
https://github.com/google/ExoPlayer/blob/d5d78238773d6dc53b54849ca10d845a4421284c/demos/main/src/main/assets/media.exolist.json
(just change any one uri, to "rtsp://..."), build the demo app, and see if it plays?

I can verify the link `rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_175k.mov` works in the demo app, let me see what else is wrong.

**Comment by warren-bank** [comment]
Well..
* the barebones app (above) proves that the RTSP library works..
* the weird thing is.. I'm walking through my app with the debugger and everything looks good.. but
  - the PlayerView is black (though the progress/seek bar shows the correct length of the video)
  - the same logcat error occurs
  - ExoPlayer times out and forces closed
* the only difference between the two (that I can think of and need to look into) is that the real app calls:
  - playWhenReady(false)
  - does some stuff in another thread
  - after that finished, calls playWhenReady(true)
  - maybe the RTSP library doesn't like that?.. TBD

**Comment by claincly** [comment]
Could you try building the player with a `MediaSource.Factory` instead of constructing a media source directly?

I.e., build your player using

```
Context context = getApplicationContext();
player =
    new SimpleExoPlayer.Builder(context)
         .setMediaSourceFactory(new DefaultMediaSourceFactory(context))
         .build();
```

And call `player.setMediaItems()` to set your `MediaItem`.

**Comment by warren-bank** [comment]
yep..
* that's exactly what the problem is

this does work:
* play an HLS video
* add the RTSP video to the queue
* click next (HLS -> RTSP)
* can prev/next back and forth between the two

this does __not__ work:
* play the RTSP video
  - because doing so causes the flow (using pseudo code):
    * playWhenReady(false)
    * await concatenatingMediaSource.clear()
    * await concatenatingMediaSource.add(rtsp)
    * playWhenReady(true)

**Comment by warren-bank** [comment]
that flow must prevent `dataLoadable.load()` from getting called

I don't know the internals, but when playWhenReady(true) is called and the RTSP stream begins to play.. it hasn't been called.. and results in a NullPointerException (as described above)

However, when the RTSP MediaSource is in a concatenatingMediaSource that's already playing (ie: playWhenReady is true) and it becomes the active item in the playlist.. all is good.

**Comment by warren-bank** [comment]
I'll make a version of the barebones app to illustrate this flow.. which should result in the same force close behavior. A tiny app would describe the problem better than I can with words..

**Comment by warren-bank** [comment]
Sorry, I crashed at about 4:30am..

Quick update:
* I did make a barebones version that uses all the same hocus-pocus as the real app
  - it still worked
* Next, I'm going to attack this from the other end
  - test the real app by removing random features one at a time until the RTSP stream plays without crashing
  - I'll report back what I find..
    * no idea right now what it could be
    * may not be relevant in a general case, may be.. TBD

**Comment by warren-bank** [comment]
[Bam!](https://github.com/warren-bank/Android-ExoPlayer-AirPlay-Receiver/commit/bbff6d3a59b990da63b3368595a6df7d5ef86ba3)

OK, so the problem was this line:
```java
  player.setRepeatMode(Player.REPEAT_MODE_ALL);
```

* initialize the player with this line, and RTSP stream immediately causes a fatal error force closing the entire app
* comment it out, and the RTSP stream plays properly
* while the RTSP stream is playing:
  - if "repeat mode" is changed by the user using the GUI, a fatal error force closes the entire app

**Comment by warren-bank** [comment]
I made a very minimal app to demonstrate the error:
* [code](https://github.com/warren-bank/Android-ExoPlayer-demos/blob/r2.14.0.001/5-rtsp-barebones-issue-8994/src/main/java/com/google/android/exoplayer2/rtspdemo_8994/MainActivity.java#L100)
* [apk](https://github.com/warren-bank/Android-ExoPlayer-demos/releases/tag/r2.14.0.001)

**Comment by ojw28** [comment]
@claincly - Note that `RtspMediaSource` needs to support creation of multiple `RtspMediaPeriod` instances, whose lifespans may overlap.

I think it might be a good idea to move toward a loading model like `ProgressiveMediaSource`, where the `MediaSource` doesn't load anything during preparation. Preparation completes immediately with a placeholder timeline. When a `ProgressiveMediaPeriod` is created later, it reports what's needed up to the parent `ProgressiveMediaSource` so that it can create and report the correct `Timeline`. This would let you move `RtspClient` instantiation down into `RtspMediaPeriod`, which I think would solve this problem.

**Comment by claincly** [comment]
Fixed in 1ca0efd.
